### ISSUE TITLE
Error when getting battery

### ISSUE THREAD

**Issue by gfaidley** [joeshaw/carwings#1]
Hi Joe,

First off thank you for creating this utility and the home kit one!

I'm wondering if something may have changed with what carwings is returning when the battery is quarried.  When I run the battery command I get the following:

```
carwings -email <email> -password <password> battery (with my user name and password of course)
Logging into Carwings...
Getting latest retrieved battery status...
ERROR: json: invalid use of ,string struct tag, trying to unmarshal "" into int
```
I'm brand new to the go language so have not dug into the source yet.  I'll try to take a look and see if there is anything obvious.

Thanks in advance
Galen

**Comment by joeshaw** [comment]
I will take a look at it soon, in the meantime you can set `Debug = true` in the source and it’ll dump out the responses from the server. That’ll be the easiest way to figure out what’s going on.

It looks like the API is sending `""` when we expect `"0"` or `null`. I suspect it’s some minor difference between our accounts.

**Comment by joeshaw** [comment]
specifically, here: https://github.com/joeshaw/carwings/blob/master/carwings.go#L39

**Comment by gfaidley** [comment]
Thanks for getting back to me.  Here are the results with the debug flag set to true:

```
Getting latest retrieved battery status...
HTTP/1.1 200 OK
Transfer-Encoding: chunked
Content-Type: application/json; charset=utf-8
Date: Sun, 04 Feb 2018 21:09:43 GMT
Server: Apache
Strict-Transport-Security: max-age=15768000

2b7
{"status":200,"BatteryStatusRecords":{"OperationResult":"START","OperationDateAndTime":"2018\/02\/04 21:08","BatteryStatus":{"BatteryChargingStatus":"NOT_CHARGING","BatteryCapacity":"12","BatteryRemainingAmount":"12","BatteryRemainingAmountWH":"","BatteryRemainingAmountkWH":""},"PluginState":"NOT_CONNECTED","CruisingRangeAcOn":"106704.0","CruisingRangeAcOff":"126464.0","TimeRequiredToFull":{"HourRequiredToFull":"0","MinutesRequiredToFull":"50"},"TimeRequiredToFull200":{"HourRequiredToFull":"0","MinutesRequiredToFull":"50"},"TimeRequiredToFull200_6kW":{"HourRequiredToFull":"0","MinutesRequiredToFull":"50"},"NotificationDateAndTime":"2018\/02\/04 21:09","TargetDate":"2018\/02\/04 21:08"}}
0



ERROR: json: invalid use of ,string struct tag, trying to unmarshal "" into int
```

Is the error on the BatteryRemainingAmountWH and BatteryRemainingAmountkWH that are returning "" ?

**Comment by joeshaw** [comment]
Should be fixed now, please give it a try.  Thank you for the report!

**Comment by gfaidley** [comment]
Thank for the quick fix.  I'm not sure what the etiquette is on commenting here or creating a new issue but there is another problem.  Looks like maybe a float to an int conversion:

```
Getting latest retrieved battery status...
HTTP/1.1 200 OK
Transfer-Encoding: chunked
Content-Type: application/json; charset=utf-8
Date: Mon, 05 Feb 2018 02:21:55 GMT
Server: Apache
Strict-Transport-Security: max-age=15768000


2b7
{"status":200,"BatteryStatusRecords":{"OperationResult":"START","OperationDateAndTime":"2018\/02\/04 22:32","BatteryStatus":{"BatteryChargingStatus":"NOT_CHARGING","BatteryCapacity":"12","BatteryRemainingAmount":"12","BatteryRemainingAmountWH":"","BatteryRemainingAmountkWH":""},"PluginState":"NOT_CONNECTED","CruisingRangeAcOn":"106704.0","CruisingRangeAcOff":"126464.0","TimeRequiredToFull":{"HourRequiredToFull":"0","MinutesRequiredToFull":"50"},"TimeRequiredToFull200":{"HourRequiredToFull":"0","MinutesRequiredToFull":"50"},"TimeRequiredToFull200_6kW":{"HourRequiredToFull":"0","MinutesRequiredToFull":"50"},"NotificationDateAndTime":"2018\/02\/04 22:32","TargetDate":"2018\/02\/04 22:32"}}
0



ERROR: json: cannot unmarshal number 106704.0 into Go struct field .CruisingRangeAcOn of type int
```

**Comment by gfaidley** [comment]
I have implemented a fix.  At first I changed those values to floats everywhere but I realized that breaks you API so casting to an int earlier seemed more appropriate.

Let me see if I can figure out pull requests.  In case I can't here is the code:

Line 509 & 510
```
		CruisingRangeACOn:  int(batrec.CruisingRangeAcOn),
		CruisingRangeACOff: int(batrec.CruisingRangeAcOff),
```

**Comment by gfaidley** [comment]
Just realized I only sent half of the change.
Lines 482 & 3

```
			CruisingRangeAcOn  float64 `json:",string"`
			CruisingRangeAcOff float64 `json:",string"`
```

**Comment by joeshaw** [comment]
Ugh, this API...

I think using `json.Number` is a better way to address it, because it will take either `float64` or `int`.  I appreciate your digging on this, I will take a look later today.

**Comment by gfaidley** [comment]
Thanks, sounds good.  This is the first I've ever even looked at Go (I hard to search for what a float was called) but that does seem more versatile.

I did find another issue too.  The battery percentage is always zero.  Looking at the code it looks like there should be a json field for that but that appears missing in what is returned from my car.  The official Nissan app does display a percentage but it seems to always change in 1/12 increments.  I assume they are calculating the percentage from the bars.

**Comment by gfaidley** [comment]
Here is a nicer formatted version of what is retuned:

```
2b7
{
    "status":200,
    "BatteryStatusRecords":
    {
        "OperationResult":"START",
        "OperationDateAndTime":"2018\/02\/04 22:32",
        
        "BatteryStatus":
        {
            "BatteryChargingStatus":"NOT_CHARGING",
            "BatteryCapacity":"12",
            "BatteryRemainingAmount":"12",
            "BatteryRemainingAmountWH":"",
            "BatteryRemainingAmountkWH":""
        },
        
        "PluginState":"NOT_CONNECTED",
        "CruisingRangeAcOn":"106704.0",
        "CruisingRangeAcOff":"126464.0",
        
        "TimeRequiredToFull":
        {
            "HourRequiredToFull":"0",
            "MinutesRequiredToFull":"50"
        },
        
        "TimeRequiredToFull200":
        {
            "HourRequiredToFull":"0",
            "MinutesRequiredToFull":"50"
        },
        
        "TimeRequiredToFull200_6kW":
        {
            "HourRequiredToFull":"0",
            "MinutesRequiredToFull":"50"
        },
        
        "NotificationDateAndTime":"2018\/02\/04 22:32",
        "TargetDate":"2018\/02\/04 22:32"
    }
}
0
```

**Comment by joeshaw** [comment]
Just out of curiosity, what region do you live in, and what model year leaf do you have?

**Comment by gfaidley** [comment]
I'm in NNA (Illinois to be exact) .  The car is a 2013 that I just had upgraded to the 3G radio (it originally had the 2G that was shut off a year ago).  I assume you have a newer one?

At first glance I seem to be getting pretty much the same responses as the "Get Cached Vehicle Status" listed here:
https://github.com/joshperry/carwings/blob/master/protocol.markdown

**Comment by gfaidley** [comment]
I had a few minutes to work on this tonight.  I just overrode the state of charge with a value calculated by dividing the  Amount Remaining by the Capacity.  If I knew anything about go or JSON processing in go I imagine it would be fairly easy to put a conditional on if that field exists.

I also I found so more information about this.  It appears as though the 2016 added the more accurate SOC.
https://github.com/jdhorne/pycarwings2/pull/22
https://github.com/jdhorne/pycarwings2/issues/13

On a side note I really wanted to thank you for pulling this together.  I've used the remote climate control more in the last week now that it is reliable and works with siri than I used it in the last 2 years.

**Comment by joeshaw** [comment]
Ah, wonderful! Thank you for digging into this.  Indeed, I have a 2017 model.

I may not be able to get to this week, but I won't let it drop off my radar.

And it's my pleasure!  I now use Homekit to turn on the heater constantly, at every chance I get 😄

**Comment by joeshaw** [comment]
Welp, I was able to sneak some time in to fix it now and it was pretty straightforward.  Let me know if you run into any more issues!

**Comment by gfaidley** [comment]
Hi Joe, I finally got a minute to work on this again tonight.  I may have misunderstood something but did you fix the percent charge issue?  I think I pulled down the latest and it seems to behave the same.  Looking through the code I also didn't see anything that would indicate it was changed (I could have missed it however).

**Comment by joeshaw** [comment]
No, I missed the percent charge issue.  I will look into that.

**Comment by gfaidley** [comment]
No worries.  Here is the code I used when I hacked it in.  I was not sure how to test if a value had come from the json file or not.  I'm also not sure how go deals with division of ints so I may have some casts I don't need.  The + 0.5 is to make this the equivalent of rounding (rather than truncating).  From my googling go just added rounding support.

`StateOfCharge:      int( float64(batrec.BatteryStatus.BatteryRemainingAmount)/float64(batrec.BatteryStatus.BatteryCapacity) * 100.0 + 0.5),`
